tests/framework/static_analysis.py

```diff
@@ -0,0 +1,618 @@
+
+            # This gets triggered in the __lockfile function on x86_64 (syscall number is loader before a branching instruction,
+            # but if the branch is not taken, linear execution will eventually hit a ret. So during backpropagation we
+            # would need to skip the section of assembly between "jmp" and "ret", but our script doesn't do anything
+            # sophisticated like that and thus instead tries to analyse this branch where the syscall number register
+            # gets clobbered, and it eventually hits a "pop" which it doesnt understand). The syscall in question is
+            # "futex", and we call that one a million time elsewhere anyway.
+            #
+            # See: https://github.com/kraj/musl/blob/ffb23aef7b5339b8c3234f4c6a93c488dc873919/src/stdio/__lockfile.c#L4
+            if isinstance(syscall_nr, Exception):
+                logger.warning(
+                    "Failed to resolve syscall number for instruction %s in function %s: %s",
+                    insn,
+                    fn.name,
+                    syscall_nr,
+                )
+                continue
+
+            syscall_name = seccomp.resolve_syscall(
+                arch.seccomp_arch, syscall_nr
+            ).decode("utf-8")
+            if syscall_name not in found_syscalls:
+                found_syscalls[syscall_name] = []
+
+            found_syscalls[syscall_name].append(
+                [None if isinstance(arg, Exception) else arg for arg in syscall_args]
+            )
+
+    return found_syscalls
+
+
+def load_seccomp_rules(seccomp_path: Path):
+    """Loads seccomp rules from the given file, and presents them as a dictionary
+    mapping syscalls to a list of individual filters. Each individual filter
+    describes some restriction of the arguments that are allowed to be passed
+    to the syscall."""
+    filters = json.loads(seccomp_path.read_text("utf-8"))
+
+    all_filters = (
+        filters["vcpu"]["filter"] + filters["vmm"]["filter"] + filters["api"]["filter"]
+    )
+    allowlist = {}
+
+    for seccomp_filter in all_filters:
+        syscall_name = seccomp_filter["syscall"]
+
+        if syscall_name not in allowlist:
+            allowlist[syscall_name] = []
+
+        allowlist[syscall_name].append(
+            {arg["index"]: arg["val"] for arg in seccomp_filter.get("args", [])}
+        )
+
+    return allowlist
+
+
+def determine_unneeded_seccomp_rules(seccomp_rules, found_syscalls):
+    """Based on the given list of syscall determined through static analysis, compute which of the
+    given seccomp rules are redundant. By 'redundant' we here mean that no syscall that would match
+    it is actually present in the given list of syscalls."""
+
+    # TODO: We could also determine "too broad" rules here: If all actual invocations of a syscall specific a parameter,
+    # but the rule does not restrict that parameter, we could recommend the strengthen the rule to specify the parameter!
```

Comment:
`s/we could recommend the strengthen the rule/we could recommend to strengthen the rule/`

Reply:
fixed!